application/F3DStarter.cxx

```diff
@@ -700,6 +708,123 @@ void F3DStarter::Render()
+    {
+      char* val = std::getenv(candidate);
+      if (val != nullptr)
+      {
+        std::filesystem::path path(val);
+        if (std::filesystem::is_directory(path))
+        {
+          return path;
+        }
+      }
+    }
+
+    return std::filesystem::current_path();
+  };
+
+  const auto applyFilenameTemplate = [&](const std::string& fn_template)
+  {
+    std::string filename = fn_template;
+
+    {
+      std::regex re("\\{app\\}");
+      filename = std::regex_replace(filename, re, F3D::AppName);
+    }
+    {
+      std::regex re("\\{version\\}");
+      filename = std::regex_replace(filename, re, F3D::AppVersion);
+    }
+    {
+      std::regex re("\\{model\\}");
+      if (std::regex_search(filename, re))
+      {
+        const auto file = this->Internals->FilesList[this->Internals->CurrentFileIndex];
+        filename = std::regex_replace(filename, re, file.filename().string());
+      }
+    }
+
+    return std::filesystem::path(filename);
+  };
+
+  const auto findAvailableFilename = [&](const std::string& idealFilename)
+  {
+    const auto appendFilenameNumber = [](const std::filesystem::path& path, size_t n)
+    {
+      std::stringstream ss;
+      const std::string fn = path.string();
+      const std::string::size_type i = fn.find_last_of('.');
+      if (i != std::string::npos)
+      {
+        ss << fn.substr(0, i) << "_" << n << fn.substr(i);
+      }
+      else
+      {
+        ss << fn << "_" << n;
+      }
+      return ss.str();
+    };
+
+    std::string fn = idealFilename;
+    for (size_t i = 2; std::filesystem::exists(fn) && i < 1e8; ++i)
+    {
+      fn = appendFilenameNumber(idealFilename, i);
+    }
+    if (std::filesystem::exists(fn))
+    {
+      throw std::runtime_error("could not find available filenamne");
+    }
+    return fn;
+  };
+
+  std::filesystem::path path = applyFilenameTemplate(filenameTemplate);
+  if (!path.is_absolute())
+  {
+    path = getScreenshotDir() / path;
+  }
+  path = findAvailableFilename(path.string());
+  std::filesystem::create_directories(std::filesystem::path(path).parent_path());
+  f3d::log::info("saving screenshot to " + path.string());
+
+  f3d::window& window = this->Internals->Engine->getWindow();
+
+  std::stringstream cameraMetadata;
+  {
+    const auto state = window.getCamera().getState();
+    const auto vec3toJson = [](const std::array<double, 3>& v)
+    {
+      std::stringstream ss;
+      ss << "[" << v[0] << ", " << v[1] << ", " << v[2] << "]";
+      return ss.str();
+    };
+    cameraMetadata << "{\n";
+    cameraMetadata << "  \"pos\": " << vec3toJson(state.pos) << ",\n";
+    cameraMetadata << "  \"foc\": " << vec3toJson(state.foc) << ",\n";
+    cameraMetadata << "  \"up\": " << vec3toJson(state.up) << ",\n";
+    cameraMetadata << "  \"angle\": " << state.angle << "\n";
+    cameraMetadata << "}\n";
+  }
+
+  window.renderToImage()
+    .setMetadata("camera", cameraMetadata.str())
```

Comment:
should we set this when using `--output` too ? that would be more coherent imo.

Reply:
could do, the templating could also be used. Would that mean moving some code to the lib or is that all in the app side still?

Reply:
`--output` is an application side option, so most of it should be in the app side.